File: app/src/main/java/com/nosfistis/mike/refene/refene/RefenesActivity.java
package com.nosfistis.mike.refene.refene;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.ActionMode;
import android.view.GestureDetector;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.View;
import android.widget.TextView;

import com.nosfistis.mike.refene.R;
import com.nosfistis.mike.refene.database.Refene;
import com.nosfistis.mike.refene.shared.RecyclerViewAdapter;
import com.nosfistis.mike.refene.viewmodel.RefeneViewModel;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.Objects;

public class RefenesActivity extends AppCompatActivity implements ActionMode.Callback {

    public static final String REFENE_ID = "refId";

    private static final int NEW_PURCHASE_REQUEST = 1;

    private RecyclerView mRecyclerView;
    private RecyclerViewAdapter mAdapter;
    private ArrayList<String> nameDataset = new ArrayList<>();
    private ArrayList<Float> totalDataset = new ArrayList<>();
    private ArrayList<Long> idDataset = new ArrayList<>();
    private ArrayList<Float> ownedDataset = new ArrayList<>();
    private float totalSum = 0;
    private long refID;
    private ActionMode actionMode;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.refenes_activity);
        mRecyclerView = findViewById(R.id.my_recycler_view);

        RefeneViewModel refeneViewModel = ViewModelProviders.of(this).get(RefeneViewModel.class);

        Intent intent = getIntent();
        refID = intent.getLongExtra(RefenesActivity.REFENE_ID, (long) -1);

        Refene refene = null;

        if (refID == -1) {
            throw new IllegalArgumentException("Refene id not provided");
        }

        refene = refeneViewModel.getRefeneById(refID);

//            List<Transaction> data = db.getAllRefeneTransactions(refID);
//            db.close();
//
//            for (Transaction transaction : data) {
//                nameDataset.add(transaction.getPerson().getName());
//                totalDataset.add(transaction.getPrice());
//                idDataset.add(transaction.getPerson().getId());
//
//                totalSum += transaction.getPrice();
//            }

        setTitle(String.valueOf(refID));

        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        mRecyclerView.setHasFixedSize(true);
        mRecyclerView.setLayoutManager(layoutManager);
        mRecyclerView.addItemDecoration(new DividerItemDecoration(mRecyclerView.getContext(), layoutManager.getOrientation()));
        GestureDetector mGestureDetector = new GestureDetector(this, new RecyclerViewOnGestureListener());
        mAdapter = new RecyclerViewAdapter(nameDataset, totalDataset, ownedDataset, mGestureDetector);
        mRecyclerView.setAdapter(mAdapter);
        mRecyclerView.addOnItemTouchListener(mAdapter);

        updateCalculations();
    }

    @Override
    protected void onResume() {
        super.onResume();
        updateRecords();
    }

    private void updateRecords() {
//        db.open();
//        List<Transaction> transactions = db.getAllRefeneTransactions(refID);
//        db.close();
//
//        totalSum = 0;

//        for (Transaction transaction : transactions) {
//            int index = idDataset.indexOf(transaction.getPerson().getId());
//            if (index == -1) {
//                nameDataset.add(transaction.getPerson().getName());
//                totalDataset.add(transaction.getPrice());
//                idDataset.add(transaction.getPerson().getId());
//            } else {
//                nameDataset.set(index, transaction.getPerson().getName());
//                totalDataset.set(index, transaction.getPrice());
//                idDataset.set(index, transaction.getPerson().getId());
//            }
//
//            totalSum += transaction.getPrice();
//        }
        updateCalculations();
    }

    private void updateCalculations() {
        float share = totalSum / nameDataset.size();
        ownedDataset.clear();
        for (float t : totalDataset) {
            ownedDataset.add(share - t);
        }
        mAdapter.notifyDataSetChanged();
        NumberFormat numberFormat = NumberFormat.getCurrencyInstance();
        ((TextView) findViewById(R.id.totalText)).setText(numberFormat.format(totalSum));
    }

    public void onAddButtonClick(View view) {
        Intent intent = new Intent(this, NewPurchaseActivity.class);
        intent.putExtra(NewPurchaseActivity.REFENE_ID, refID);
        startActivityForResult(intent, NEW_PURCHASE_REQUEST);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == NEW_PURCHASE_REQUEST && resultCode == RESULT_OK) {
            updateRecords();
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_refenes, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean onCreateActionMode(ActionMode actionMode, Menu menu) {
        MenuInflater inflater = actionMode.getMenuInflater();
        inflater.inflate(R.menu.context_menu, menu);
        return true;
    }

    @Override
    public boolean onPrepareActionMode(ActionMode actionMode, Menu menu) {
        return false;
    }

    @Override
    public void onDestroyActionMode(ActionMode actionMode) {
        this.actionMode = null;
    }

    @Override
    public boolean onActionItemClicked(ActionMode actionMode, MenuItem menuItem) {
        switch (menuItem.getItemId()) {
            // TODO: Edit refenes name
            case R.id.context_menu_edit:
                actionMode.finish();
                return true;
            // Delete selected items from list and database
            case R.id.context_menu_delete:
//                db.open();
//                List<Integer> selections = mAdapter.getSelectedItems();
//                for (int i : selections) {
//                    db.removePersonFromRefene(idDataset.get(i), refID);
//                    totalSum -= totalDataset.get(i);
//                    mAdapter.removeData(i);
//                    idDataset.remove(i);
//                    updateCalculations();
//                }
//                db.close();
                actionMode.finish();
                return true;
            default:
                return false;
        }
    }

    private class RecyclerViewOnGestureListener extends GestureDetector.SimpleOnGestureListener {
        @Override
        public boolean onSingleTapConfirmed(MotionEvent e) {
            View view = mRecyclerView.findChildViewUnder(e.getX(), e.getY());
            if (view != null) {
                if (actionMode != null) {
                    mAdapter.toggleSelection(mRecyclerView.getChildAdapterPosition(view));
                    return super.onSingleTapConfirmed(e);
                }
                TextView nameView = view.findViewById(R.id.nameText);
                Intent intent = new Intent(view.getContext(), PersonalBidsActivity.class);
                intent.putExtra("person", nameView.getText().toString());
                intent.putExtra("id", idDataset.get(mRecyclerView.getChildAdapterPosition(view)));
                intent.putExtra(RefenesActivity.REFENE_ID, refID);
                startActivity(intent);
            }
            return super.onSingleTapConfirmed(e);
        }

        public void onLongPress(MotionEvent e) {
            if (actionMode != null) {
                return;
            }

            actionMode = startActionMode(RefenesActivity.this);

            View view = mRecyclerView.findChildViewUnder(e.getX(), e.getY());
            mAdapter.toggleSelection(mRecyclerView.getChildAdapterPosition(Objects.requireNonNull(view)));

            super.onLongPress(e);
        }
    }
}

File: app/src/main/java/com/nosfistis/refene/RefenesActivity.kt
package com.nosfistis.refene

import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.ActionMode
import android.view.GestureDetector
import android.view.Menu
import android.view.MenuItem
import android.view.MotionEvent
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import java.text.NumberFormat

class RefenesActivity : AppCompatActivity(), ActionMode.Callback {
    private var mRecyclerView: RecyclerView? = null
    private var mAdapter: RecyclerViewAdapter? = null
    private val nameDataset = ArrayList<String?>()
    private val totalDataset = ArrayList<Float>()
    private val idDataset = ArrayList<Long>()
    private val ownedDataset = ArrayList<Float>()
    private var totalSum = 0f
    private var refID: Long = 0
    private var actionMode: ActionMode? = null
    private var db: DatabaseHandler? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.refenes_activity)
        mRecyclerView = findViewById(R.id.my_recycler_view)

        db = DatabaseHandler(this)

        refID = intent.getLongExtra("refID", -1)

        if (refID != -1L) {
            db!!.open()
            val data: List<Transaction> =
                db!!.getAllRefeneTransactions(refID)
            db!!.close()

            for (transaction in data) {
                val person = transaction.getPerson()
                if (person != null) {
                    nameDataset.add(person.name)
                    idDataset.add(person.id)
                }
                totalDataset.add(transaction.price)

                totalSum += transaction.price
            }
        } else {
            db!!.open()
            refID = db!!.addRefene()
            db!!.close()
        }

        title = refID.toString()

        val layoutManager = LinearLayoutManager(this)
        mRecyclerView!!.setHasFixedSize(true)
        mRecyclerView!!.setLayoutManager(layoutManager)
        mRecyclerView!!.addItemDecoration(
            DividerItemDecoration(
                mRecyclerView!!.context,
                layoutManager.orientation
            )
        )
        val mGestureDetector = GestureDetector(this, RecyclerViewOnGestureListener())
        mAdapter = RecyclerViewAdapter(nameDataset, totalDataset, ownedDataset, mGestureDetector)
        mRecyclerView!!.setAdapter(mAdapter)
        mRecyclerView!!.addOnItemTouchListener(mAdapter!!)

        updateCalculations()
    }

    override fun onResume() {
        super.onResume()
        updateRecords()
    }

    private fun updateRecords() {
        db!!.open()
        val transactions: List<Transaction> =
            db!!.getAllRefeneTransactions(refID)
        db!!.close()

        totalSum = 0f

        for (transaction in transactions) {
            val index = idDataset.indexOf(transaction.getPerson()!!.id)
            if (index == -1) {
                nameDataset.add(transaction.getPerson()!!.name)
                totalDataset.add(transaction.price)
                idDataset.add(transaction.getPerson()!!.id)
            } else {
                nameDataset[index] = transaction.getPerson()!!.name
                totalDataset[index] = transaction.price
                idDataset[index] = transaction.getPerson()!!.id
            }

            totalSum += transaction.price
        }
        updateCalculations()
    }

    private fun updateCalculations() {
        val share = totalSum / nameDataset.size
        ownedDataset.clear()
        for (t in totalDataset) {
            ownedDataset.add(share - t)
        }
        mAdapter!!.notifyDataSetChanged()
        val numberFormat = NumberFormat.getCurrencyInstance()
        (findViewById<TextView>(R.id.totalText)!!).text = numberFormat.format(totalSum.toDouble())
    }

    fun onAddButtonClick(view: View?) {
        val intent: Intent = Intent(this, NewPurchaseActivity::class.java)
        intent.putExtra("refID", refID)
        startActivityForResult(intent, NEW_PURCHASE_REQUEST)
    }

    @Deprecated("This method has been deprecated in favor of using the Activity Result API\n      which brings increased type safety via an {@link ActivityResultContract} and the prebuilt\n      contracts for common intents available in\n      {@link androidx.activity.result.contract.ActivityResultContracts}, provides hooks for\n      testing, and allow receiving results in separate, testable classes independent from your\n      activity. Use\n      {@link #registerForActivityResult(ActivityResultContract, ActivityResultCallback)}\n      with the appropriate {@link ActivityResultContract} and handling the result in the\n      {@link ActivityResultCallback#onActivityResult(Object) callback}.")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (requestCode == NEW_PURCHASE_REQUEST && resultCode == RESULT_OK) {
            updateRecords()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_refenes, menu)
        return true
    }

    override fun onCreateActionMode(actionMode: ActionMode, menu: Menu): Boolean {
        val inflater = actionMode.menuInflater
        inflater.inflate(R.menu.context_menu, menu)
        return true
    }

    override fun onPrepareActionMode(actionMode: ActionMode, menu: Menu): Boolean {
        return false
    }

    override fun onDestroyActionMode(actionMode: ActionMode) {
        this.actionMode = null
    }

    override fun onActionItemClicked(actionMode: ActionMode, menuItem: MenuItem): Boolean {
        when (menuItem.itemId) {
            R.id.context_menu_edit -> {
                actionMode.finish()
                return true
            }

            R.id.context_menu_delete -> {
                db!!.open()
                val selections: List<Int> = mAdapter!!.getSelectedItems()
                for (i in selections) {
                    db!!.removePersonFromRefene(idDataset[i], refID)
                    totalSum -= totalDataset[i]
                    mAdapter!!.removeData(i)
                    idDataset.removeAt(i)
                    updateCalculations()
                }
                db!!.close()
                actionMode.finish()
                return true
            }

            else -> return false
        }
    }

    private inner class RecyclerViewOnGestureListener : GestureDetector.SimpleOnGestureListener() {
        override fun onSingleTapConfirmed(e: MotionEvent): Boolean {
            val view = mRecyclerView!!.findChildViewUnder(e.x, e.y)
            if (view != null) {
                if (actionMode != null) {
                    mAdapter!!.toggleSelection(mRecyclerView!!.getChildAdapterPosition(view))
                    return super.onSingleTapConfirmed(e)
                }
                val nameView = view.findViewById<View>(R.id.nameText) as TextView
                val intent = Intent(view.context, PersonalBidsActivity::class.java)
                intent.putExtra("person", nameView.text.toString())
                intent.putExtra("id", idDataset[mRecyclerView!!.getChildAdapterPosition(view)])
                intent.putExtra("refID", refID)
                startActivity(intent)
            }
            return super.onSingleTapConfirmed(e)
        }

        override fun onLongPress(e: MotionEvent) {
            if (actionMode != null) {
                return
            }

            actionMode = startActionMode(this@RefenesActivity)

            val view = mRecyclerView!!.findChildViewUnder(e.x, e.y)

            if (view != null) {
                mAdapter!!.toggleSelection(mRecyclerView!!.getChildAdapterPosition(view))
            }

            super.onLongPress(e)
        }
    }

    companion object {
        private const val NEW_PURCHASE_REQUEST = 1
    }
}
